llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp

```diff
@@ -879,6 +879,16 @@ Instruction *InstCombinerImpl::visitFMul(BinaryOperator &I) {
     if (Constant *NegC = ConstantFoldUnaryOpOperand(Instruction::FNeg, C, DL))
       return BinaryOperator::CreateFMulFMF(X, NegC, &I);
 
+  if (I.hasNoNaNs() && I.hasNoSignedZeros()) {
+    // (uitofp bool X) * Y --> X ? Y : 0
+    // Y * (uitofp bool X) --> X ? Y : 0
+    // Note INF * 0 is NaN.
+    if (match(Op0, m_UIToFP(m_Value(X))) && X->getType()->isIntOrIntVectorTy(1))
+      return SelectInst::Create(X, Op1, ConstantFP::get(I.getType(), 0.0));
+    if (match(Op1, m_UIToFP(m_Value(X))) && X->getType()->isIntOrIntVectorTy(1))
+      return SelectInst::Create(X, Op0, ConstantFP::get(I.getType(), 0.0));
+  }
+
```

Comment:
You don't need nsz if you also introduce a copysign:

https://alive2.llvm.org/ce/z/Be9G_s

Reply:
Not obvious to me if the form with copysign is better or not, it's 2 instructions -> 2 instructions

Reply:
With copysign we need to check that `uitofp` is one-use.

Reply:
Is it alright to just start with this case?

Reply:
Yes